### PR TITLE
Fix alloc-dealloc mismatches

### LINKED ISSUE #54637: Allocation/Deallocation mismatch for MethodTable::MethodData in CoreCLR
As part of https://github.com/dotnet/runtime/pull/54580, I've discovered an alloc/decalloc mismatch in CoreCLR.

We sometimes allocate space for `MethodTable::MethodData` objects with `new[]` instead of `new`, but always delete with `delete`.

Allocation:

https://github.com/dotnet/runtime/blob/0416c3469b7bccff087b89db3d9967dfd36070a3/src/coreclr/vm/methodtable.cpp#L8544-L8563

Dealloc:

https://github.com/dotnet/runtime/blob/0416c3469b7bccff087b89db3d9967dfd36070a3/src/coreclr/vm/methodtable.cpp#L7901-L7912

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Use specialized new operators to correctly allocate space with the non-array new allocator for MethodDataObject and MethodDataInterfaceImpl.

Fix alloc/dealloc mismatch in ILStubResolver by just using `new` instead of new-ing up a buffer and placement new-ing into it.

Fixes #54637

### CODE REVIEW

File: src/coreclr/vm/ilstubresolver.cpp [hunk 1]
```diff
@@ -344,7 +344,7 @@ ILStubResolver::AllocGeneratedIL(
     if (!UseLoaderHeap())
     {
         NewArrayHolder<BYTE>             pNewILCodeBuffer = new BYTE[cbCode];
-        NewArrayHolder<CompileTimeState> pNewCompileTimeState = (CompileTimeState*)new BYTE[sizeof(CompileTimeState)];
+        NewHolder<CompileTimeState>      pNewCompileTimeState = new CompileTimeState{};
```
Comment (jkotas) [inline src/coreclr/vm/ilstubresolver.cpp:347]: For my education - does `new CompileTimeState{}` allocate the space without calling any constructor?

**Review by jkotas — APPROVED** [review]
Thanks!

File: src/coreclr/vm/ilstubresolver.cpp (hunk 1 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/ilstubresolver.cpp:347]: Good catch. It actually zero-inits, so I can remove the `memcpy` on the next line.

File: src/coreclr/vm/ilstubresolver.cpp [hunk 2]
```diff
@@ -344,7 +344,7 @@ ILStubResolver::AllocGeneratedIL(
...
         memset(pNewCompileTimeState, 0, sizeof(CompileTimeState));
```
Comment (jkoritzinsky) [inline src/coreclr/vm/ilstubresolver.cpp:348]:
```suggestion
```

**Comment by jkoritzinsky** [comment]
Test failure is #54713

**Pull request merged**